### PR TITLE
AIP-84 Migrate views /object/historical_metrics_data to Fast API

### LINKED ISSUE #42623: AIP-84 Migrate /object/historical_metrics_data from views to FastAPI
### Description

Migrate views `/object/historical_metrics_data` to Fast API.

### Use case/motivation

[AIP-84](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-84+UI+REST+API)

### Related issues

Relates to: https://github.com/apache/airflow/issues/42366

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by bugraoz93**
Closes: #42623

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/api_fastapi/views/ui/objects.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+from airflow.api_fastapi.db.common import get_session
+from airflow.api_fastapi.views.router import AirflowRouter
+from airflow.utils import timezone
+
+object_router = AirflowRouter(tags=["Object"])
+
+
+@object_router.get("/object/historical_metrics_data", include_in_schema=False)
```
Comment (bbovenzi) [inline airflow/api_fastapi/views/ui/objects.py]: I didn't like the `/object/` nomenclature in `views.py`. That was merely an attempt to mark which endpoints returned objects instead of FAB views.

Let's take the opportunity to give a more descriptive name. In this case, we are using the historical metrics for a dashboard view. So let's make it a `dashboard_router`

Comment (pierrejeambrun) [reply airflow/api_fastapi/views/ui/objects.py]: Good idea. The final route could be `ui/dashboard/historical_metrics_data`. So we could put that in `views/ui/dashboard.py`

File: airflow/api_fastapi/views/ui/objects.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    :param date_to_check: the string value to be parsed
+    :param allow_empty: Set True to return none if empty str or None
+    :param strict: if False, it will fall back on the dateutil parser if unable to parse with pendulum
+    """
+    if allow_empty is True and not date_to_check:
+        return None
+    try:
+        return timezone.parse(date_to_check, strict=strict)
+    except (TypeError, ParserError):
+        abort(400, f"Invalid datetime: {date_to_check!r}")
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/views/ui/objects.py]: I think should be part of a commong `SafeDatetimeQuery` in `parameters.py`. I think there is an integration with `AfterValidatior` of pydantic to call custom validation function no query params easily.

So anybody that want's to get a safe date from query parameter can just re-use that with:
```
async def historical_metrics_data(
    start_date: SafeDateTimeQuery,
    end_date: SafeDateTimeQuery,
    session: Annotated[Session, Depends(get_session)],
) -> dict:
```

Or maybe just `DateTimeQuery` if we plan to make all of the 'date' query param safe by default.

Comment (pierrejeambrun) [reply airflow/api_fastapi/views/ui/objects.py]: You can get inspiration from there https://github.com/fastapi/fastapi/discussions/10131

File: airflow/api_fastapi/views/ui/objects.py (hunk 1 above)
Comment (bugraoz93) [reply airflow/api_fastapi/views/ui/objects.py]: Make sense! I made the changes that were needed. Thanks for the both! 

Indeed, we should take this as an advantage to shape them properly. Please make all these comments because I am first doing lift and shift to see if it would be enough. Please also include if you need any additional things from the UI end.

File: airflow/api_fastapi/views/ui/objects.py (hunk 2 above)
Comment (bugraoz93) [reply airflow/api_fastapi/views/ui/objects.py]: Awesome one! Thanks! 

Adding this method there bugged me out too. I was going to ask what is strategy be for generalising this but I went offline quickly after pushing my last commit. You read my mind and answered my question :)

I have included the method and `AfterValidator` in `parameters.py` and added them as type definitions to the dashboard endpoint. The method is different from other Param classes but at the end, this is a parameter Annotation and included the method there as well. I am not sure if we should create a `common` or `utils` for `parameter.py` for this kind of case. How it's looking now?

**Comment by bbovenzi** [comment]
One nit after looking at the autogenerated code. Let's drop the word `data`. Just `historical_metrics`

File: airflow/ui/openapi-gen/requests/types.gen.ts [hunk 3]
```diff
@@ -96,6 +96,15 @@ export type NextRunAssetsResponse = {
   [key: string]: unknown;
 };
 
+export type HistoricalMetricsDataData = {
+  endDate: string;
+  startDate: string;
+};
+
+export type HistoricalMetricsDataResponse = {
+  [key: string]: unknown;
```
Comment (bbovenzi) [inline airflow/ui/openapi-gen/requests/types.gen.ts]: @pierrejeambrun Is there a way we can better interpret the response type here?

In `v1-generated` it is just `object` which leads to `unknown` over here.

File: airflow/api_fastapi/parameters.py [hunk 4]
```diff
@@ -235,6 +239,24 @@ def depends(self, last_dag_run_state: DagRunState | None = None) -> _LastDagRunS
...
+    try:
+        return timezone.parse(date_to_check, strict=True)
+    except (TypeError, ParserError):
+        raise HTTPException(
+            400, f"Invalid datetime: {date_to_check!r}. Please check the date parameter have this value."
+        )
+
+
+# Common Safe DateTime
+DateTimeQuery = Annotated[str, AfterValidator(_safe_parse_datetime)]
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/parameters.py:259]: Nice

**Review by pierrejeambrun — COMMENTED** [review]
Looking good, just the comment above for the response schema and then we can merge I think

File: airflow/ui/openapi-gen/requests/types.gen.ts (hunk 3 above)
Comment (pierrejeambrun) [reply airflow/ui/openapi-gen/requests/types.gen.ts]: Good question.

Indeed, most of the `private/ui/object` endpoints were only used by the UI and therefore did not have proper schema definition (for documentation purpose) and are just returning a plain 'dict' without any information which end up in the spec as a generic 'object'. This isn't ideal.

To better leverage the type system we could simply define a PydanticModel that defines the return type of the endpoint, and therefore doc and codegen would be accurate. I think we need to do that for all of the UI endpoints that we are migrating to have a cleaner interface. That will remove the need for manually duplicating types in the front-end like we used to do for private endpoints.
```typescript
interface HistoricalMetricsData {
  dagRunStates: {
    [K in CamelCase<RunState>]: number;
  };
  dagRunTypes: {
    [K in CamelCase<DagRun["runType"]>]: number;
  };
  taskInstanceStates: {
    [K in TaskState extends string ? CamelCase<K> : never]: number;
  };
}
```

So basically we need a:

```python
class HistoricalMetricsResponse(BaseModel):
    ....
```

Comment (bbovenzi) [reply airflow/ui/openapi-gen/requests/types.gen.ts]: Yes, that was a main motivation for AIP-84 in the first place, we didn't have good typing for any of our UI endpoints and manually writing them in the UI could break very easily.

Comment (bugraoz93) [reply airflow/ui/openapi-gen/requests/types.gen.ts]: Great one! I agree, we should take the opportunity to make everything as descriptive as possible so that UI can be more consistent. I will include a model for all UI endpoints in the future. 
I have included the models for this MR. I have included every detail rather than `Dict[str, int]` to be more descriptive, I hope it works.

**Review by pierrejeambrun — APPROVED** [review]

**Comment by pierrejeambrun** [comment]
@bugraoz93 can you rebase the branch before I merge please, it is quite behind main. Thanks :)

**Comment by bugraoz93** [comment]
For sure, @pierrejeambrun, I rebased it. Thanks! :)

**Comment by bugraoz93** [comment]
Okay, merges after my rebase caused conflicts. Solving them and rebasing again :)

**Comment by bbovenzi** [comment]
Looks like we need another rebase and then running pre-commit for the auto-generated files.

**Comment by bugraoz93** [comment]
Rebased and run the hooks.

**Comment by bugraoz93** [comment]
While running the hooks, a new PR merged. I rebased again just in case. The speed of the development :rocket: :)

**Review by pierrejeambrun — APPROVED** [review]
